pkg/framework/aries/default.go

```diff
@@ -122,11 +150,102 @@ func setAdditionalDefaultOpts(frameworkOpts *Aries) error {
 		frameworkOpts.msgSvcProvider = &noOpMessageServiceProvider{}
 	}
 
-	// TODO add SecretLock creation here.. #1150
+	return nil
+}
+
+func createMasterKeyAndSecretLock(opts *Aries) error {
+	keySize := sha256.Size
+
+	passphrase, salt, err := getPassphraseAndSalt(opts, keySize)
+	if err != nil {
+		return err
+	}
+
+	masterLocker, err := hkdf.NewMasterLock(passphrase, sha256.New, salt)
+	if err != nil {
+		return err
+	}
+
+	file, err := os.OpenFile(defMKPath, os.O_WRONLY|os.O_CREATE, 0600)
+	if err != nil {
+		return err
+	}
+
+	fileName := file.Name()
+
+	fStat, err := file.Stat()
+	if err != nil {
+		return err
+	}
+
+	if fStat.Size() < int64(base64.URLEncoding.EncodedLen(keySize)) {
+		// if file content doesn't contain a valid key size, generate new key, encrypt and overwrite file
+		err = generateAndWriteMK(file, keySize, masterLocker)
+		if err != nil {
+			return err
+		}
+	}
+
+	err = file.Close()
+	if err != nil {
+		return err
+	}
+
+	r, err := local.MasterKeyFromPath(fileName)
+	if err != nil {
+		return err
+	}
+
+	s, err := local.NewService(r, masterLocker)
+	if err != nil {
+		return err
+	}
+
+	opts.secretLock = s
+
+	return nil
+}
+
+func generateAndWriteMK(file io.Writer, keySize int, masterLocker secretlock.Service) error {
+	masterKeyContent := random.GetRandomBytes(uint32(keySize))
+
+	masterLockEnc, err := masterLocker.Encrypt("", &secretlock.EncryptRequest{
+		Plaintext: string(masterKeyContent)})
+	if err != nil {
+		return err
+	}
+
+	_, err = file.Write([]byte(masterLockEnc.Ciphertext))
+	if err != nil {
+		return err
+	}
 
 	return nil
 }
 
+func getPassphraseAndSalt(opts *Aries, keySize int) (string, []byte, error) {
+	passphrase := "secretPwd"
+	if opts.passphrase != "" {
+		passphrase = opts.passphrase
+		opts.passphrase = ""
+	}
+
+	salt := make([]byte, keySize)
+	if opts.salt != "" {
+		if len(opts.salt) != keySize {
+			return "", nil, fmt.Errorf("invalid salt option provided")
+		}
+
+		copy(salt, opts.salt)
+		opts.salt = ""
+	} else {
+		// default -  no salt
+		salt = nil
```

Comment:
Why not generate a new random salt?

Reply:
salt is optional, if the user doesn't provide one, the framework shouldn't generate one.

salt and passphrase are not stored anywhere... they are the `key` to unlock the master key which can be stored either in a file or an env variable.. for the default behaviour, the framework will generate a master key and secure it in a file so that future New() calls can still read it (of course the user must pass in the same passphrase and salt options they provided the first time to unlock the master key)

Reply:
also notice that `passphrase` and `salt` options are cleared once used to avoid keep in them in memory
see lines 230 and 240 above

Reply:
salt and passphrase are removed from framework.go and default.go